Airgap existing cluster proposal (#710)
replicatedhq/kots: design/airgap-existing-cluster.md

# Support for installing KOTS Admin Console in Airgapped Clusters

## Goals

- Install kotsadm to an existing cluster where the cluster does not have internet access
- Install kotsadm to an existing cluster from a workstation where the workstation does not have internet access at the time of installation
- Support different credentials for push / pull (workstation / cluster) to the local registry
- Support temporary/dynamic and persistent/static credentials for pushing kotsadm images to the registry

> - Support only persistent/static credentials for pulling kotsadm images to the registry

Reviewer: Are these credentials really needed?  All kotsadm images are public, but also they will be downloaded as a github asset according to this doc.

Reviewer 2: should this say "pushing" to the registry? I think that's how I had read it

Reviewer 3: The problem is just that the enterprise isn't going to have a registry that is open, so we expect they'll need a password to fetch the images, even if the images are otherwise public. But also just realized this is phrased wrong. The password isn't necessary to pull images **to** the registry, but **from** the local registry.

Reviewer 3: ```suggestion
- Support only persistent/static credentials for pulling kotsadm images from the airgapped registry
```

- Support different credentials for pushing vs pulling images to the local registry

## Non Goals

- Create a single installer for kotsadm + application
- Create a single download for kotsadm, kots plugin, application and license
- Automated installs of kotsadm to existing clusters in airgap mode
- Support for installing versions of KOTS different from the KOTS CLI version installed

## Background

It's not currently possible to install KOTS Admin Console into an existing cluster that doesn't have outbound internet access.

## High-Level Design

Airgap Workstation: the computer with kubectl access to the airgap cluster
Public Workstation: the computer with access to the internet

From the public workstation:
1. The user downloads the kots plugin suitable for Airgap Workstation.

> 1. The user downloads the kotsadm archive containing kots container images

Reviewer: I'm not clear on this.  Does the user download the archive from github or does the "kots install" command download everything for the user?

Reviewer 3: We were vague there. :) Worst case, the vendor would send an email to the admin with a link to download. They might add a redirector or cname but it's a straight download. It is **not** downloaded via `kots install`.

It's conceivable to combine downloading the kots plugin and kotsadm archive into the same download tarball, and that would probably be a better experience, but it's not a requirement. Heck, it would be better to download all three things (those two plus the application airgapped bundle) in a single download. But we're trying to limit scope here to a minimal iteration with a passable workflow. We could probably enhance that later.

1. The user downloads the airgap bundle of the application
1. The user copies the kots plugin, the kotsadm archive and the application airgap bundle onto a medium (eg a USB stick) that can be transfered into the airgapped network.

From the airgap workstation:

> 1. The user will install the kots plugin 

Reviewer: Is this a manual `cp` command or do we provide a script to do this?

Reviewer 3: To capture: We talked and agreed to have an install script like kurl airgapped install.

1. The user will install kotsadm to the cluster using the kots plugin via `kots install` passing local registry information which will be used to generate the kotsadm yaml and image pull secrets (if neccessary)
1. The user will install the application and license using the existing airgap installation process 

## Detailed Design

The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.
The files SHOULD be stored in the archive in the same format as application airgap files are stored.

Add a `kots admin-console push-images` to the kots cli to retag and push all kotsadm images to an authenticated registry that is accessible from the cluster, using the desired push credentials.
This new command is used to copy all of the KOTS Admin Console images from one location to another.

Example:

```
kubectl kots admin-console push-images ./kotsadm-1.16.1.tar.gz registry.somebigbank.com/my-app \

>   --registry-username push-user \

Reviewer: I propose we replace `registry-user` and `registry-password` with a `registry-secret-name`.   The secret will contain a `.dockerconfigjson` key and the json will contain a key for every registry used by the process (the push registry and the pull registry).  This is the standard k8s way to define registry secrets and this gets us automation at no extra cost.

Reviewer 3: Talked with @divolgin and we can skip this for now. If we hit the stretch goal above of trying to pull anonymously and if rejected then try with `~/.docker/config.json` credentials, then someone could `docker login` before hand and this would work without leaking credentials into bash history.

# Support for installing KOTS Admin Console in Airgapped Clusters

## Goals

- Install kotsadm to an existing cluster where the cluster does not have internet access
- Install kotsadm to an existing cluster from a workstation where the workstation does not have internet access at the time of installation
- Support different credentials for push / pull (workstation / cluster) to the local registry

> - Support temporary/dynamic and persistent/static credentials for pushing kotsadm images to the registry

Reviewer 2: I know it would be great to support a wide variety of credential methods, but I think if we wanted to we could cut a little scope here and have any registry push situation require passing a username and password via environment or CLI flags. Support for other fanciers auth methods could come later.

Reviewer 3: To be clear, we're not really doing anything too special to support this other than asking for the push credentials separately from the pull credentials, and only using the push credentials once. If it required more work, then we could see about cutting it from scope, but as is, it was a natural part of the design anyway, so no additional work, yet still beneficial.

- Support only persistent/static credentials for pulling kotsadm images to the registry

> - Support different credentials for pushing vs pulling images to the local registry

Reviewer 2: This seems redundant with 

```
- Support different credentials for push / pull (workstation / cluster) to the local registry
```

is there a difference between these?

Reviewer 3: Yeah, there's a difference and this part is necessary, although we can make it clearer. Pushing is a one-time action, so you can use temporary credentials to do it, but pulling can happen any time later, as Kubernetes needs to pull in images, so the credentials need to be long-lived. But even with that, it's possible to design a system such that both credentials need to be the same because we only accept the credentials once. We consciously designed it such that you could specify different credentials.

## Non Goals

- Create a single installer for kotsadm + application
- Create a single download for kotsadm, kots plugin, application and license

> - Automated installs of kotsadm to existing clusters in airgap mode

Reviewer 2: Could we clarify "Automated" Here?

```suggestion
- Headless/Automated installs of kotsadm to existing clusters in airgap mode
```

Reviewer 3: How about "Unattended"?

Reviewer 3: ```suggestion
- Unattended, automatic installs of kotsadm to existing clusters in airgap mode
```

> - Support private (authenticated) local hosting of kotsadm images

Reviewer 2: I'm not sure what we mean by "local hosting" but this needs to include the ability to store the kotsadm images in my private registry

Reviewer 3: I think that was a leftover from an earlier iteration.
```suggestion
- Support private (authenticated) local hosting of kotsadm images
```

- Support for installing versions of KOTS different from the KOTS CLI version installed

## Background

It's not currently possible to install KOTS Admin Console into an existing cluster that doesn't have outbound internet access.

## High-Level Design

Airgap Workstation: the computer with kubectl access to the airgap cluster
Public Workstation: the computer with access to the internet

From the public workstation:
1. The user downloads the kots plugin suitable for Airgap Workstation.
1. The user downloads the kotsadm archive containing kots container images
1. The user downloads the airgap bundle of the application
1. The user copies the kots plugin, the kotsadm archive and the application airgap bundle onto a medium (eg a USB stick) that can be transfered into the airgapped network.

From the airgap workstation:
1. The user will install the kots plugin 

> 1. The user will copy the container images in the kotsadm archive to the local registry using the kots kubectl plugin

Reviewer 2: ```suggestion
1. The user will run a `kubectl kots` command to copy the container images in the kotsadm archive to the local registry, providing registry details for pushing
```

1. The user will install kotsadm to the cluster using the kots plugin via `kots install` passing local registry information which will be used to generate the kotsadm yaml and image pull secrets (if neccessary)

> 1. The user will install thte application and license using the existing airgap installation process 

Reviewer 2: ```suggestion
1. The user will install the application and license using the existing airgap installation process 
```

## Detailed Design

The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.

> The files SHOULD be stored in the archive in the same format as application airgap files are stored.

Reviewer 2: wish we had that format documented somewhere

Reviewer: i think this refers to image format.  if that's the case, the answer is, `docker-archive`, not `OCI`.

> The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
> The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.
> The files SHOULD be stored in the archive in the same format as application airgap files are stored.
> 
> Add a `kots admin-console push-images` to the kots cli to retag and push all kotsadm images to an authenticated registry that is accessible from the cluster, using the desired push credentials.
> This new command is used to copy all of the KOTS Admin Console images from one location to another.

Reviewer 3: ```suggestion
* The CI process, on tag, will create the kotsadm archive and include it in the GitHub Release.
* The filename will be `kots_airgap_<version>.tar.gz` and this file contains .tar file exports of all kotsadm images.
* The files SHOULD be stored in the archive in the same format as application airgap files are stored.

* Add a `kots admin-console push-images` to the kots cli to retag and push all kotsadm images to an authenticated registry that is accessible from the cluster, using the desired push credentials.
* This new command is used to copy all of the KOTS Admin Console images from one location to another.
```

Example:

```
kubectl kots admin-console push-images ./kotsadm-1.16.1.tar.gz registry.somebigbank.com/my-app \
  --registry-username push-user \
  --registry-password abcdef

> ```

Reviewer 2: lgtm

> `--registry-username` and `--registry-password` are both optional.

Reviewer 3: ```suggestion
Stretch: `--registry-username` and `--registry-password` are both optional.
```

> If not present, attempt to authenticate using the already authenticated local docker credentials.

Reviewer 2: As noted above, I would vote this goal be a `Stretch` as well, as docker credential providers can get weird to integrate from what I understand, they involve forking out to other binaries sometimes

Reviewer 3: Good point. I'm fine with making it a stretch.
```suggestion
Stretch: If not present, attempt to authenticate using the already authenticated local docker credentials.
```

Reviewer 3: Although if we don't do that stretch, then what's the behavior without it? Are the credentials required then?

Reviewer: When credentials are not provided, we should assume the registry is anonymous.  If it is not, the command will fail with an appropriate error message.

Reviewer 3: @divolgin said on a call that we already have logic for handling trying the registry anonymously and then trying `.docker/config.json` creds if needed, so sounds like this might be easy to do. Again, it can still be a stretch as it's not strictly necessary.

Stretch: if still not authenticated, prompt for username/password, use these to push but do not save locally.
Stretch: if username is specified and password is not, prompt for password to use, do not save locally.

KOTS will then load the images from the file in the first parameter, retag them to the local registry and push. 
All images will retain the same image name and tag, just the registry, namespace will change.

At this point, the images are in the registry, but not running in the cluster yet.
A workflow could scan these images before continuing.

Next, add a flag to `kots install` to collect / rewrite local registry endpoint/namespace into kots manifests, using the desired pull credentials.
This will assume that the images are already in the destination.

For example:

```
kubectl kots install --kotsadm-registry registry.somebigbank.com/my-app \
  --registry-username pull-user \
  --registry-password fedcba \
  app-name
```

> If a `registry-username` and `registry-password` are provided, an `imagePullSecret` will be created and deployed with the registry credentials and added to all kotsadm pods.

Reviewer 2: I think we also need to consider the case where no registry-username/password are provided, but the user wants to direct kots to add an already-existing imagePullSecret to all workloads

Reviewer 3: @dexhorthy Hmm... how likely is that? Can we start without it? We briefly discussed using advanced secret management like short-lived Vault credentials but decided it was out of scope. If the existing imagePullSecret is a simple username and password, then having to replicate it isn't horrible at least, although it's not DRY and may miss regular rotating if there's some existing mechanism to rotate creds. I think we can cut that for now.

The only difference between the command above and an online installation is that the `image` tag will be rewritten in all KOTS Admin Console manifests to point to `registry.somebigbank.com/my-app` and an `imagePullSecret` MAY be added.
All image names and tags will be unchanged.
For example, the kotsadm-operator will be `registry.somebigbank.com/my-app/kotsadm-operator:v1.16.1`.

Stretch: Additionally, a new `--yaml` flag will be added to have `kots` only generate the yaml (not apply) to support gitops workflows.

## Limitations

- No support for custom branding in airgap installations to existing clusters
- No support for `requireMinimalRBACPrivileges` set to `true` (all installs will be cluster admin, which matches all use cases of KOTS install today)

## Assumptions

> - The workstation will have kubectl access to the cluster

Reviewer 2: should these specify which workstation?

Reviewer 3: ```suggestion
- The airgap workstation will have kubectl access to the cluster
```

> - The workstation must have push access to an image registry that the cluster has pull access 

Reviewer 3: ```suggestion
- The airgap workstation must have push access to an image registry that the cluster has pull access 
```
